### PR TITLE
Adapt PowderreduceP2d to usage with SNAP data

### LINKED ISSUE #35943: Adapt PowderreduceP2d to usage with SNAP data
**Is your feature request related to a problem? Please describe.**
During my PhD work i have developed the data reduction workflow algorithm PowderReduceP2d (already oart of mantid) for the soon to be build TOF powder diffractometer POWTEX. As the instrument is not running yet, I have extended the algorithm to work with measuring data from other instruments, in this case the SNAP instrument located at the SNS in Oak Ridge.

**Describe the solution you'd like**
The code is already finished and I have submitted a pull request some weeks ago. I would be thankful if someone can review my changes and hope this issue brings some attention to the changes. 

**Additional context**
I also hope that I can get some labels added to the issue and the pull request. Diffraction and Powder are the ones fitting best I think.

### PR DESCRIPTION
**Pull request by YannickMeinerzhagen**
**Description of work**
The Workflow algorithm PowderreduceP2d has been modified for several reasons.
1. As the algorithm AlignDetectors was deprecated, it was replaced with the algorithm ApplyDiffCal.
2. A small Bug for the ResetNegatives2D function was fixed. 
3. The algorithm was modified to be applicable to data collected at the instrument SNAP at SNS.
4. The possibility to use a vanadium background file was added.
5. Parameter centerWavelength was removed.

**Purpose of work**
The algorithm PowderReduceP2d was developed for the creation of multidimensional neutron time of flight datasets. These datasets are necessary for the new multidimensional Rietveld refinement method that is being developed for the diffractometer POWTEX to be build at FRM-II in Garching. For more information on the multidimensional Rietveld refinement look at:

https://doi.org/10.1107/S1600576723002819

https://doi.org/10.1107/S1600576717005398

https://doi.org/10.1107/S1600576715016520

As the instrument POWTEX is not build up yet, we are testing the reduction and refinement process with data from different instruments. The extension to other instruments is interesting for others as well, as the method can be applied to all TOF instruments with large area detectors.

**Summary of work**
1. The algorithm was replaced as was recommended.
2. The resetting of negative values to a set value was done by creating a new workspace. Not setting the attribute distribution to True resulted in an error, so that was added to the function.
3. The SNAP instrument uses a different calibration method than was used before (detcal files instead of cal files), which made a different calibration routine necessary.
4. The instrument POWGEN has started using background measurements for the vanadium correction, so this has been added as an optional functionality. 
5. The parameter center wavelength was removed because it was unnecessary. 

**Further detail of work**
3. To reduce data collected at the instrument SNAP it was first necessary to implement the usage of detcal files for the calibration of the detectors. Depending on the fileextension (cal or detcal) the code automatically follows the corresponding calibration path. To make the usage of masking files easier, the missing option to use xml files via the loadMask algorithm has been implemented.

**To do:**

- [x] Check if test function for this algorithm still works
- [x] If necessary, adapt testing data
- [x] Check .rst file for any changes necessary
- [x] If necessary, make changes to flow chart
- [x] Add release notes

**To test:**
Example for testing can be found in the test function associated with this algorithm. If available, POWGEN or SNAP data should generally work but some fiddling with initial parameters might be necessary.

Closes #35943.

At this moment the base of this PR is main, if this is not correct please let me know. A fitting label for this PR would be diffraction, unfortunately I cannot add a label to the PR or I have not found the option to do this yet.

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: docs/source/diagrams/PowderReduceP2D_ProcessCanRun.dot [hunk 1]
```diff
@@ -24,6 +26,10 @@ digraph PowderReduceP2D_ProcessCanRun {
 
   subgraph decisions{
     $decision_style
+    maskFile        [label="Is masking file specified?"]
+    detcal        [label="Is calibration file of type detcal?"]
+    detcal2        [label="Is calibration file of type detcal?"]
+    detcal3        [label="Is calibration file of type detcal?"]
     edgebinning     [label="Is edgebinning specified?"]
```
Comment (robertapplin) [inline docs/source/diagrams/PowderReduceP2D_ProcessCanRun.dot:33]:
```suggestion
    $decision_style
    maskFile        [label="Is masking file specified?"]
    detcal          [label="Is calibration file of type detcal?"]
    detcal2         [label="Is calibration file of type detcal?"]
    detcal3         [label="Is calibration file of type detcal?"]
    edgebinning     [label="Is edgebinning specified?"]
```

File: docs/source/diagrams/PowderReduceP2D_ProcessCanRun.dot [hunk 2]
```diff
@@ -32,13 +38,23 @@ digraph PowderReduceP2D_ProcessCanRun {
...
   }
 
-  CanData	                -> FindDetectorsPar
+  SampleData	            -> maskFile
+  maskFile   	            -> LoadMask  [label="Yes"]
+  maskFile   	            -> detcal  [label="No"]
+  LoadMask                  -> detcal
+  detcal                    -> LoadDetcal  [label="Yes"]
+  detcal                    -> FindDetectorsPar  [label="No"]
+  LoadDetcal                    -> FindDetectorsPar
```
Comment (robertapplin) [inline docs/source/diagrams/PowderReduceP2D_ProcessCanRun.dot:47]:
```suggestion
  detcal                    -> FindDetectorsPar  [label="No"]
  LoadDetcal                -> FindDetectorsPar
```

File: docs/source/diagrams/PowderReduceP2D_ProcessVanadiumRun.dot [hunk 3]
```diff
@@ -28,6 +30,10 @@ digraph PowderReduceP2D_ProcessVanRun {
 
   subgraph decisions{
     $decision_style
+    maskFile        [label="Is masking file specified?"]
+    detcal        [label="Is calibration file of type detcal?"]
+    detcal2        [label="Is calibration file of type detcal?"]
+    detcal3        [label="Is calibration file of type detcal?"]
     edgebinning     [label="Is edgebinning specified?"]
```
Comment (robertapplin) [inline docs/source/diagrams/PowderReduceP2D_ProcessVanadiumRun.dot:37]:
```suggestion
    maskFile        [label="Is masking file specified?"]
    detcal          [label="Is calibration file of type detcal?"]
    detcal2         [label="Is calibration file of type detcal?"]
    detcal3         [label="Is calibration file of type detcal?"]
    edgebinning     [label="Is edgebinning specified?"]
```

File: docs/source/diagrams/PowderReduceP2D_ProcessVanadiumRun.dot [hunk 4]
```diff
@@ -36,13 +42,24 @@ digraph PowderReduceP2D_ProcessVanRun {
...
   }
 
-  VanadiumData	            -> FindDetectorsPar
+  SampleData	            -> maskFile
+  maskFile   	            -> LoadMask  [label="Yes"]
+  maskFile   	            -> detcal  [label="No"]
+  LoadMask                  -> detcal
+  detcal                    -> LoadDetcal  [label="Yes"]
+  detcal                    -> FindDetectorsPar  [label="No"]
+  LoadDetcal                    -> FindDetectorsPar
```
Comment (robertapplin) [inline docs/source/diagrams/PowderReduceP2D_ProcessVanadiumRun.dot:51]:
```suggestion
  detcal                    -> FindDetectorsPar  [label="No"]
  LoadDetcal                -> FindDetectorsPar
```

File: docs/source/release/v6.8.0/Diffraction/Powder/New_features/35872.rst [hunk 5]
```diff
@@ -0,0 +1 @@
+-:ref:`PowderReduceP2D <algm-PowderReduceP2D>` has been adapted to work with SNAP measuring data
```
Comment (robertapplin) [inline docs/source/release/v6.8.0/Diffraction/Powder/New_features/35872.rst:1]:
```suggestion
- :ref:`PowderReduceP2D <algm-PowderReduceP2D>` has been adapted to work with SNAP measuring data.
```

**Review by robertapplin — CHANGES_REQUESTED** [review]
Just had a quick look through. I won't be able to do a full review, so thought I'd just look through the code. I spotted a couple of misaligned indentations - apologies for how picky this is! Apart from that, the code looks good

**Comment by YannickMeinerzhagen** [comment]
Thanks a lot for your input. I dont mind picky stuff, it has to be done and will only quicken the following review :)

**Comment by MialLewis** [comment]
Hi Yannick, You're currently occupying all 8 of our linux builders with builds triggered as a result of your pushes to your remote branch associate with this PR. I don't suppose you could fix your PR locally before pushing? Let me know if you have any queries regarding how to do this.

**Comment by YannickMeinerzhagen** [comment]
I'm sorry, didn't realise that would happen when I just accept the change requests here! I will do changes locally next time, thanks for letting me know.

**Comment by MialLewis** [comment]
No worries, I think if you click `Add suggestion to batch` you can commit all the recommended changes as one commit, which will only trigger one build.

**Review by robertapplin — APPROVED** [review]
Thanks for the changes. The code looks good and it works as explained 👍

**Comment by gemmaguest** [comment]
> I'm sorry, didn't realise that would happen when I just accept the change requests here! I will do changes locally next time, thanks for letting me know.

@YannickMeinerzhagen You can also temporarily close a PR if you want to push to it without kicking off the builds - just re-open it again when you want to run the builds. I haven't tried applying suggested changes when it's closed but I imagine it should still work.

**Pull request merged**